client/app/react-components/QueryEditor.js

```diff
@@ -0,0 +1,228 @@
+    canExecuteQuery: PropTypes.bool.isRequired,
+    executeQuery: PropTypes.func.isRequired,
+    saveQuery: PropTypes.func.isRequired,
+    updateQuery: PropTypes.func.isRequired,
+    listenForResize: PropTypes.func.isRequired,
+    queryExecuting: PropTypes.bool.isRequired,
+  }
+
+  static defaultProps = {
+    syntax: 'sql',
+    autocompleteQuery: false,
+    dataSource: { options: { doc: '' } },
+    dataSources: [],
+  }
+
+  constructor(props) {
+    super(props);
+    this.refEditor = React.createRef();
+    this.state = {
+      schema: null, // eslint-disable-line react/no-unused-state
+      keywords: [], // eslint-disable-line react/no-unused-state
+      autocompleteQuery: false,
+      queryText: this.props.queryText,
+    };
+    const self = this;
+    langTools.addCompleter({
+      getCompletions(state, session, pos, prefix, callback) {
+        if (prefix.length === 0) {
+          callback(null, []);
+          return;
+        }
+        callback(null, self.state.keywords);
+      },
+    });
+    this.onPaste = (text) => {
+      const editor = self.refEditor.editor;
+      editor.session.doc.replace(editor.selection.getRange(), text);
+      const range = editor.selection.getRange();
+      window.setTimeout(() => {
+        editor.selection.setRange(range);
+      }, 0);
+    };
+
+    this.onLoad = (editor) => {
+      // Release Cmd/Ctrl+L to the browser
+      editor.commands.bindKey('Cmd+L', null);
+      editor.commands.bindKey('Ctrl+L', null);
+
+      //   self.props.QuerySnippet.query((snippets) => {
+      //     const snippetManager = snippetsModule.snippetManager;
+      //     const m = {
+      //       snippetText: '',
+      //     };
+      //     m.snippets = snippetManager.parseSnippetFile(m.snippetText);
+      //     snippets.forEach((snippet) => {
+      //       m.snippets.push(snippet.getSnippet());
+      //     });
+      //     snippetManager.register(m.snippets || [], m.scope);
+      //   });
+      editor.focus();
+      self.props.listenForResize((e) => { console.log(e); editor.resize(); });
+    };
+
+    this.formatQuery = () => {
+      this.props.formatQuery(this.props.dataSource.syntax, this.state.queryText)
+        .then((queryText) => { this.props.updateQuery(queryText); this.setState({ queryText }); })
+        .catch(error => this.props.toastr.error(error));
+    };
+  }
+
+  static getDerivedStateFromProps(nextProps, prevState) {
+    if (!nextProps.schema.fulfilled) {
+      return { ...prevState, keywords: [], autocompleteQuery: false };
+    } else if (nextProps.schema.value.schema !== prevState.schema) {
+      const schema = nextProps.schema.value.schema;
+      return {
+        ...prevState,
+        schema,
+        keywords: buildKeywordsFromSchema(schema),
+        autocompleteQuery: (schema.reduce((totalLength, table) =>
+          totalLength + table.columns.length, 0) <= 5000 && nextProps.autocompleteQuery),
+      };
+    }
+    return prevState;
```

Comment:
https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops

"It should return an object to update the state, or null to update nothing."